cabal-install/Distribution/Client/IndexUtils.hs

```diff
@@ -147,16 +177,62 @@ indexBaseName repo = repoLocalDir repo </> fn
 --
 -- This is a higher level wrapper used internally in cabal-install.
 getSourcePackages :: Verbosity -> RepoContext -> IO SourcePackageDb
-getSourcePackages verbosity repoCtxt | null (repoContextRepos repoCtxt) = do
-  warn verbosity $ "No remote package servers have been specified. Usually "
-                ++ "you would have one specified in the config file."
-  return SourcePackageDb {
-    packageIndex       = mempty,
-    packagePreferences = mempty
-  }
-getSourcePackages verbosity repoCtxt = do
-  info verbosity "Reading available packages..."
-  pkgss <- mapM (\r -> readRepoIndex verbosity repoCtxt r) (repoContextRepos repoCtxt)
+getSourcePackages verbosity repoCtxt =
+    getSourcePackagesWithState verbosity repoCtxt IndexStateHead
+
+-- | Variant of 'getSourcePackages' which allows to request a global 'IndexState'
```

Comment:
-- Variant of 'getSourcePackages' which allows getting the source packages at a particular 'IndexState'.
    --
    -- Current choices are either the latest, or the index as it was at a particular time.
    -- TODO: ... etc